File: backend/app/src/main/kotlin/net/animeta/backend/controller/v2/RecordPostsController.kt
package net.animeta.backend.controller.v2

import net.animeta.backend.dto.PostDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.repository.HistoryRepository
import net.animeta.backend.repository.RecordRepository
import net.animeta.backend.serializer.PostSerializer
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/v2/records/{id}/posts")
class RecordPostsController(val recordRepository: RecordRepository,
                            val historyRepository: HistoryRepository,
                            val postSerializer: PostSerializer) {
    data class GetResponse(val posts: List<PostDTO>)

    @GetMapping
    fun get(@PathVariable id: Int,
            @RequestParam(required = false) options: PostSerializer.Options?): GetResponse {
        val record = recordRepository.findById(id).orElseThrow { ApiException.notFound() }
        val posts = record.histories
        return GetResponse(posts.map { postSerializer.serialize(it, options ?: PostSerializer.legacyOptions()) })
    }
}

File: backend-ts/src/controllers/record_posts.controller.ts
import { Controller, Get, Param, ParseIntPipe, Query } from "@nestjs/common";
import { PostDTO } from 'shared/types';
import { ApiException } from "./exceptions";
import { RecordService } from "src/services/record.service";
import { Repository } from "typeorm";
import { InjectRepository } from "@nestjs/typeorm";
import { PostSerializer, PostSerializerOptions } from "src/serializers/post.serializer";
import { History } from 'src/entities/history.entity'

@Controller('/api/v4/records/:id/posts')
export class RecordPostsController {
  constructor(
    private recordService: RecordService,
    @InjectRepository(History) private historyRepository: Repository<History>,
    private postSerializer: PostSerializer,
  ) {}

  @Get()
  async get(
    @Param('id', new ParseIntPipe()) id: number,
    @Query('options') optionsJson: string | null,
  ): Promise<{ posts: PostDTO[] }> {
    const record = await this.recordService.get(id)
    if (!record) throw ApiException.notFound()
    const posts = await this.historyRepository.find({
      where: { record },
      order: { id: 'DESC' },
    })
    const options: PostSerializerOptions =
      optionsJson ? JSON.parse(optionsJson) : PostSerializer.legacyOptions()
    return {
      posts: await Promise.all(posts.map(it => this.postSerializer.serialize(it, options))),
    }
  }
}
